core/types/transaction.go

```
@@ -287,6 +293,28 @@ func (tx *Transaction) To() *common.Address {
 	return copyAddressPtr(tx.inner.to())
 }
 
+// IsL1MessageTx returns true if the transaction is an L1 cross-domain tx.
+func (tx *Transaction) IsL1MessageTx() bool {
+	return tx.Type() == L1MessageTxType
+}
+
+// AsL1MessageTx casts the tx into an L1 cross-domain tx.
+func (tx *Transaction) AsL1MessageTx() *L1MessageTx {
+	if !tx.IsL1MessageTx() {
+		return nil
+	}
+	return tx.inner.(*L1MessageTx)
+}
+
+// L1MessageQueueIndex returns the L1 queue index if `tx` is of type `L1MessageTx`.
+// It returns 0 otherwise.
+func (tx *Transaction) L1MessageQueueIndex() uint64 {
+	if !tx.IsL1MessageTx() {
+		return 0
```

The `QueueIndex` can be 0 on L1. wonder if returning 0 here would conflict with the actual Queue index

---

This is only used like this:

```go
if tx.IsL1MessageTx() {
	nonce = tx.L1MessageQueueIndex()
}
```

The caller should only use this value if `tx` is an `L1MessageTx`. Do you think we should panic here instead of returning 0?

---

No need to panic but just to be careful. I'm good as long as you confirm it won't cause issues at other places.

---

I think it's fine since we don't plan to use this function anywhere else, only in the trace API.